### PR TITLE
Updates to tinerary backend

### PR DESCRIPTION
**Pull request by ethangraham2001**
Added preference-based filtering of itineraries
- `Tag` type (`String`) that represents an itinerary tag, with an `object` containing common instances
- updated `Itinerary` with a list of tags, as well as upvotes / downvotes. Added a scoring method that scores an itinerary based on preferences, as well as overall rating
- implemented preference-based filtering, and sorting based on scoring in `MapViewModel`
- Added a function for computing distance between two `Location`s (goal is to use for sorting in the future)
- Added some no-argument constructors which are needed by firebase
- did some general refactoring for prettier code uwu

### CODE REVIEW

File: app/src/main/java/com/github/wanderwise_inc/app/MainActivity.kt
Comment (YofiY) [inline app/src/main/java/com/github/wanderwise_inc/app/MainActivity.kt:1]: lgtm

File: app/src/main/java/com/github/wanderwise_inc/app/model/location/Itinerary.kt [hunk 1]
```diff
@@ -72,6 +90,16 @@ data class Itinerary(
             return this
         }
 
+        /**
+         *
+         */
+        fun addTag(tag: Tag): Builder {
+            if (tags.size >= 3)
```
Comment (YofiY) [inline app/src/main/java/com/github/wanderwise_inc/app/model/location/Itinerary.kt:97]: may want to use the constant MAX_TAGS

File: app/src/main/java/com/github/wanderwise_inc/app/model/location/Location.kt [hunk 2]
```diff
@@ -20,14 +30,54 @@ data class Location(
...
+        return earthRadius * c
+    }
+
+    /**
+     * @return the distance to the start of an itinerary
+     */
+    fun distToStart(itinerary: Itinerary): Kilometers {
+        return this.distTo(itinerary.locations[0])
+    }
+
```
Comment (YofiY) [inline app/src/main/java/com/github/wanderwise_inc/app/model/location/Location.kt:72]: the math seems correct !

File: app/src/main/java/com/github/wanderwise_inc/app/model/location/Tag.kt [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+typealias Tag = String
+
+/**
+ * @brief itinerary tags used for filtering based on user preferences
+ *
+ * String representations should consist of a single word and start with a Capital letter
+ */
+object ItineraryTags {
```
Comment (YofiY) [inline app/src/main/java/com/github/wanderwise_inc/app/model/location/Tag.kt:22]: We will have to discuss further the exact tags but now for the logic, it looks good to me

File: app/src/main/java/com/github/wanderwise_inc/app/ui/map/PreviewItineraryScreen.kt [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        modifier = Modifier
+            .background(MaterialTheme.colorScheme.background)
+            .height(200.dp)
+            .clip(RoundedCornerShape(16.dp)),
+        contentAlignment = Alignment.Center,
+    ) {
+        Column (
+            horizontalAlignment = Alignment.CenterHorizontally
+        ) {
+            Text(
```
Comment (YofiY) [inline app/src/main/java/com/github/wanderwise_inc/app/ui/map/PreviewItineraryScreen.kt:84]: from the last meeting we want the itinerary to show the display the following :
- La liste des points wander stops
	- Description pour chaque stop 
	- Photos
- Cout
- Temps
- Rating
- Profile du créateur 
- Tags
We can discuss this further during our next standup meeting

Comment (YofiY) [reply app/src/main/java/com/github/wanderwise_inc/app/ui/map/PreviewItineraryScreen.kt:84]: I think that these should be attributes of the Itinerary object

**Review by ThomasPFMuller — CHANGES_REQUESTED** [review]
Ambiguities:
- Upvotes/downvotes not yet implemented?

Problems:
- fun scoreFromPreferences in Itinerary can have division by zero if there are no upvotes or downvotes -> multiplication by infinity

File: app/src/main/java/com/github/wanderwise_inc/app/viewmodel/MapViewModel.kt [hunk 5]
```diff
@@ -71,10 +71,32 @@ class MapViewModel : ViewModel() {
         }
     }
 
+    /**
+     * @param preferences user query preferences
+     * @return a list of itineraries matching a user's query preferences
+     */
+    fun getItinerariesFromPreferences(preferences: ItineraryPreferences): Flow<List<Itinerary>>{
```
Comment (YofiY) [inline app/src/main/java/com/github/wanderwise_inc/app/viewmodel/MapViewModel.kt:78]: lgtm ! good stuff

**Review by YofiY — COMMENTED** [review]
- The logic implemented for the sorting of itinerary by preferences lgtm 
- The mapViewModel follows want is needed in order to sort
- The itinerary may need additional attributes  structure modified in order to order to fit in the descriptions in the itinerary preivew banner better but it looks good for now as this is a futur task

**Review by ThomasPFMuller — APPROVED** [review]
good stuff, nice changes

**Pull request merged**